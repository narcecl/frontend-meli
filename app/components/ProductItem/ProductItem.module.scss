.MeliProductItem{
    + .MeliProductItem{
        border-top: 1px solid #eee;
    }

    padding: 16px 0;

    a{
        picture{
            width: 180px;
            height: 180px;
            position: relative;
            border-radius: 4px;
            overflow: hidden;
            display: flex;
            justify-content: center;
            align-content: center;
            flex-flow: wrap;

            img{
                width: 100%;
                height: 100%;
                object-fit: contain;
            }
        }
    }

    &__content{
        width: calc(100% - 180px);
        padding: 16px 0;

        h2{
            color: #333;
            font-weight: 300;
            line-height: 1.3;
        }
    }
}