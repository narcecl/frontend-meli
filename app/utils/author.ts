export const author = {
    name: 'Nicol√°s',
    lastname: 'Arce',
}

export default author;
