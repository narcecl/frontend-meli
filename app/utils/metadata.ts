interface MetadataProps {
    title: string;
    description?: string;
}

const defaultDescription = 'Compre productos con Env√≠o Gratis en el d√≠a en Mercado Libre Chile. Encuentre miles de marcas y productos a precios incre√≠bles.';

export function getMetadata(props: MetadataProps){
    const {
        title,
        description = defaultDescription
    } = props;

    return {
        title: `${title} | MercadoLibre üì¶`,
        description: description,
    };
}
