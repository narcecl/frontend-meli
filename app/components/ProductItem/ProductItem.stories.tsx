import type { Meta, StoryObj } from '@storybook/react';
import { ProductItem } from './ProductItem';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction
const meta: Meta<typeof ProductItem> = {
    title: 'Core/ProductItem',
    component: ProductItem,
    tags: ['autodocs'],
    argTypes: {
        id: {
            description: 'Defines Mercado Libre product ID.',
        },
        title: {
            description: 'Defines product title.'
        },
        price: {
            description: 'Product price information.'
        },
        picture: {
            description: 'Defines product thumbnail image.'
        },
        condition: {
            description: 'Defines product condition.'
        },
        free_shipping: {
            description: 'Defines if product has free shipping option.'
        },
        address: {
            description: 'Defines product address.'
        }
    }
};

export default meta;
type Story = StoryObj<typeof ProductItem>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
    args: {
        id: 'MLA1149502467',
        title: 'Figura De Acci√≥n  Tanjiro Kamado 49010 De Funko Pop! Animation',
        price: { currency: 'ARS', amount: 3399, decimals: 0 },
        picture: 'http://http2.mlstatic.com/D_744432-MLA52222036808_102022-I.jpg',
        condition: 'Nuevo',
        free_shipping: false,
        address: 'Capital Federal'
    },
};