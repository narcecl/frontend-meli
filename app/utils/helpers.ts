import { ProductConditionsProps, ProductItemProps, ProductDetailsProps } from '@/app/types/ProductItem.types';
import { CategoryProps } from '@/app/types/Category.types';

const API_PRODUCT_URL = 'https://api.mercadolibre.com/sites/MLA';
const API_ITEM_URL = 'https://api.mercadolibre.com/items';
const API_CATEGORY_URL = 'https://api.mercadolibre.com/categories/';

export async function getProducts(query: string): Promise<any> {
    return await fetch(`${API_PRODUCT_URL}/search?q=${query}&limit=4`)
    .then(res => {
        if( res.status !== 200 ) return false;
        return res.json();
    })
    .then(data => data);
}

export async function getCategoryInfo(categoryId: string): Promise<any> {
    return await await fetch(`${API_CATEGORY_URL}/${categoryId}`)
    .then(res => {
        if( res.status !== 200 ) return false;
        return res.json();
    })
    .then(data => data);
}

export async function getProductInfo(ID: string): Promise<any> {
    return await fetch(`${API_ITEM_URL}/${ID}`)
    .then(res => {
        if( res.status !== 200 ) return false;
        return res.json();
    })
    .then(data => data);
}

export async function getProductDescription (ID: string): Promise<any> {
    return await fetch(`${API_ITEM_URL}/${ID}/description`)
    .then(res => {
        if( res.status !== 200 ) return false;
        return res.json();
    })
    .then(data => data);
}

export function checkCondition(condition: string): string {
    // Return translated product condition
    if( !condition ) return '';

    let availableConditions: ProductConditionsProps[] = [
        { condition: 'new', label: 'Nuevo' },
        { condition: 'used', label: 'Usado' },
        { condition: 'refurbished', label: 'Reacondicionado' },
        { condition: 'not_specified', label: 'Estado sin especificar' },
    ];

    let foundCondition = availableConditions.find(item => item.condition === condition);
    if( foundCondition ) return foundCondition.label;
    return 'Estado sin especificar';
}

export function formatCurrency(number: number): string {
    // Format number to currency (CLP)
    const formatter = new Intl.NumberFormat('es-CL', { style: 'currency', currency: 'CLP' });
    return formatter.format(number);
}

export function getDecimals(price: number): number{
    // Get decimals from a number
    return +price.toString().split('.')[1] || 0;
};

export function getFormattedProductSearch(product: any): ProductItemProps {
    // Format product object (result) into required format
    return {
        id: product.id,
        title: product.title,
        price: {
            currency: product.currency_id,
            amount: Math.trunc(product.price),
            decimals: getDecimals(product.price),
        },
        picture: product.thumbnail,
        condition: checkCondition(product.condition),
        free_shipping: product.shipping.free_shipping,
        address: product?.address?.state_name,
    };
};

export function getFormattedProductView(product: any, description: string = ''): ProductDetailsProps {
    // Format product object into required format
    return {
        ...getFormattedProductSearch(product),
        picture: product.pictures[0]?.url || product.thumbnail,
        sold_quantity: product.sold_quantity,
        description: description,
    };
};

export function getCategoryIdByFilters(filters: CategoryProps[]): string {
    // Get the filters and return the category ID with most results 
    const categories = filters.find(filter => filter.id === 'category');
    if( !categories ) return '';

    const { values } = categories;
    const sortedCategories = values.reduce((max, current) => current.results > max.results ? current : max);
    return sortedCategories.id;
};

export function getCategoriesFromFilters(filters: any): string[] {
    // Return array with only the name's category
    const { path_from_root } = filters;
    return path_from_root.map((category: any) => category.name);
}

export async function getCategoriesFromAvailableFilters(availableFilters: any): Promise<string[]> {
    // Get the available filters and get the the category with most results
    let categoryId = getCategoryIdByFilters(availableFilters);
    let response = await getCategoryInfo(categoryId);
    let categories = getCategoriesFromFilters(response);
    return categories;
}

export async function getBreadcrumb(filters: any, available_filters: any): Promise<string[]> {
    // Return array of categories for breadcrumb, based on filters or available filters
    return filters.length ?
        getCategoriesFromFilters(filters[0].values[0]) :
        await getCategoriesFromAvailableFilters(available_filters);
}

export async function getBreadcrumbByCategory(categoryId: string): Promise<string[]> {
    // Return array of categories for breadcrumb, based on the single product category
    let response = await getCategoryInfo(categoryId);
    let categories = getCategoriesFromFilters(response);
    return categories;
}
