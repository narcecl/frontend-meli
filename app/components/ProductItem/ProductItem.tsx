import React from 'react';
import Link from "next/link";
import styles from './ProductItem.module.scss';
import { ProductItemProps } from '../../types/ProductItem.types';
import { formatCurrency } from '../../utils/helpers';
import freeShipping from '../../../assets/images/ic_shipping@2x.png';

export const ProductItem = (props: ProductItemProps) => {
    const { id, title, price, picture, free_shipping, address } = props;

    return (
        <div className={styles.MeliProductItem}>
            <div className="d-flex gap-16">
                <Link href={`items/${id}`} className="d-block">
                    <picture>
                        <img src={picture} alt={title} width={18} height={18} />
                    </picture>
                </Link>
                <div className={styles.MeliProductItem__content}>
                    <div className="w-100 w-sm-85 mb-32">
                        <div className="d-flex align-items-center justify-content-between">
                            <div className="d-flex align-items-center gap-8">
                                <span className="title--medium text--default d-block">
                                    { formatCurrency(price.amount) }
                                </span>
                                {
                                    free_shipping &&
                                    <picture>
                                        <img
                                            src={freeShipping.src}
                                            alt="Despacho gratis"
                                            title="Despacho gratis"
                                            width={18}
                                        />
                                    </picture>
                                }
                            </div>
                            <p className="body--small text--muted">{ address }</p>
                        </div>
                    </div>
                    <h2 className="body--medium text--default w-100 w-sm-50">
                        <Link href={`items/${id}`} className="d-inline-block">
                            { title }
                        </Link>
                    </h2>
                </div>
            </div>
        </div>
    );
}

export default ProductItem;