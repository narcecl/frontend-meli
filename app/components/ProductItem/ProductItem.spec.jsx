import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { formatCurrency } from '../../utils/helpers';
import ProductItem from './ProductItem';

const placeholderData = {
	id: 'MLA1149502467',
	title: 'Figura De AcciÃ³n Tanjiro Kamado 49010 De Funko Pop! Animation',
	price: { currency: 'ARS', amount: 3399, decimals: 0 },
	picture: 'http://http2.mlstatic.com/D_744432-MLA52222036808_102022-I.jpg',
	condition: 'Nuevo',
	free_shipping: false,
	address: 'Capital Federal',
};

describe('ProductItem', () => {
	it('Default render', () => {
		// Render the component
		const { container, asFragment } = render(
			<ProductItem {...placeholderData} />
		);

		// Check first render
		const firstRender = asFragment();
        expect(firstRender).toMatchSnapshot();

		// Check if the title and link
		const title = container.querySelector('h2');
		expect(title).toBeInTheDocument();
		expect(title).toHaveTextContent(placeholderData.title);
		const titleLink = title.querySelector('a');
		expect(titleLink).toBeInTheDocument();
		expect(titleLink).toHaveAttribute('href', `items/${placeholderData.id}`);

		// Check the price is rendered with correct format
		const formattedPrice = formatCurrency(placeholderData.price.amount);
		const priceElement = screen.getByText(formattedPrice);
		expect(priceElement).toBeInTheDocument();

		// Check the address is rendered
		const addressElement = screen.getByText(placeholderData.address);
		expect(addressElement).toBeInTheDocument();

		// Check the picture is rendered with correct source and alt
		const pictureElement = screen.getByRole('img', { name: placeholderData.title });
		expect(pictureElement).toBeInTheDocument();
		expect(pictureElement).toHaveAttribute('src', placeholderData.picture);
		expect(pictureElement).toHaveAttribute('alt', placeholderData.title);

		// Check the free shipping icon is not rendered
		const freeShippingIcon = screen.queryByAltText('Despacho gratis');
		expect(freeShippingIcon).not.toBeInTheDocument();
	});

	it('Render with free shipping prop', () => {
		// Render the component
		const { container, asFragment } = render(
			<ProductItem
				{...placeholderData}
				free_shipping={true}
			/>
		);

		// Check first render
		const firstRender = asFragment();
        expect(firstRender).toMatchSnapshot();

		// Check if the title and link
		const title = container.querySelector('h2');
		expect(title).toBeInTheDocument();
		expect(title).toHaveTextContent(placeholderData.title);
		const titleLink = title.querySelector('a');
		expect(titleLink).toBeInTheDocument();
		expect(titleLink).toHaveAttribute('href', `items/${placeholderData.id}`);

		// Check the price is rendered with correct format
		const formattedPrice = formatCurrency(placeholderData.price.amount);
		const priceElement = screen.getByText(formattedPrice);
		expect(priceElement).toBeInTheDocument();

		// Check the address is rendered
		const addressElement = screen.getByText(placeholderData.address);
		expect(addressElement).toBeInTheDocument();

		// Check the picture is rendered with correct source and alt
		const pictureElement = screen.getByRole('img', { name: placeholderData.title });
		expect(pictureElement).toBeInTheDocument();
		expect(pictureElement).toHaveAttribute('src', placeholderData.picture);
		expect(pictureElement).toHaveAttribute('alt', placeholderData.title);

		// Check the free shipping icon is not rendered
		const freeShippingIcon = screen.queryByAltText('Despacho gratis');
		expect(freeShippingIcon).toBeInTheDocument();
	});
});
