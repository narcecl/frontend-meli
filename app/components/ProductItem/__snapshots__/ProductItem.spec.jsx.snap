// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem Default render 1`] = `
<DocumentFragment>
  <div
    class="MeliProductItem"
  >
    <div
      class="d-flex gap-16"
    >
      <a
        class="d-block"
        href="items/MLA1149502467"
      >
        <picture>
          <img
            alt="Figura De Acci贸n Tanjiro Kamado 49010 De Funko Pop! Animation"
            height="18"
            src="http://http2.mlstatic.com/D_744432-MLA52222036808_102022-I.jpg"
            width="18"
          />
        </picture>
      </a>
      <div
        class="MeliProductItem__content"
      >
        <div
          class="w-100 w-sm-85 mb-32"
        >
          <div
            class="d-flex align-items-center justify-content-between"
          >
            <div
              class="d-flex align-items-center gap-8"
            >
              <span
                class="title--medium text--default d-block"
              >
                $3.399
              </span>
            </div>
            <p
              class="body--small text--muted"
            >
              Capital Federal
            </p>
          </div>
        </div>
        <h2
          class="body--medium text--default w-100 w-sm-50"
        >
          <a
            class="d-inline-block"
            href="items/MLA1149502467"
          >
            Figura De Acci贸n Tanjiro Kamado 49010 De Funko Pop! Animation
          </a>
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProductItem Render with free shipping prop 1`] = `
<DocumentFragment>
  <div
    class="MeliProductItem"
  >
    <div
      class="d-flex gap-16"
    >
      <a
        class="d-block"
        href="items/MLA1149502467"
      >
        <picture>
          <img
            alt="Figura De Acci贸n Tanjiro Kamado 49010 De Funko Pop! Animation"
            height="18"
            src="http://http2.mlstatic.com/D_744432-MLA52222036808_102022-I.jpg"
            width="18"
          />
        </picture>
      </a>
      <div
        class="MeliProductItem__content"
      >
        <div
          class="w-100 w-sm-85 mb-32"
        >
          <div
            class="d-flex align-items-center justify-content-between"
          >
            <div
              class="d-flex align-items-center gap-8"
            >
              <span
                class="title--medium text--default d-block"
              >
                $3.399
              </span>
              <picture>
                <img
                  alt="Despacho gratis"
                  title="Despacho gratis"
                  width="18"
                />
              </picture>
            </div>
            <p
              class="body--small text--muted"
            >
              Capital Federal
            </p>
          </div>
        </div>
        <h2
          class="body--medium text--default w-100 w-sm-50"
        >
          <a
            class="d-inline-block"
            href="items/MLA1149502467"
          >
            Figura De Acci贸n Tanjiro Kamado 49010 De Funko Pop! Animation
          </a>
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
